gef.py

```diff
@@ -5223,6 +5221,192 @@ def do_invoke(self, args: List[str]) -> None:
+                self[structure.name] = structure
+            return
+
+        def __load(self) -> ModuleType:
+            """Load a custom module, and return it."""
+            fpath = self.path
+            spec = importlib.util.spec_from_file_location(fpath.stem, fpath)
+            module = importlib.util.module_from_spec(spec)
+            sys.modules[fpath.stem] = module
+            spec.loader.exec_module(module)
+            return module
+
+        def __str__(self) -> str:
+            return self.name
+
+        def __iter__(self) -> Generator[str, None, None]:
+            _invalid = {"BigEndianStructure", "LittleEndianStructure", "Structure"}
+            _structs = {x for x in dir(self.raw) \
+                             if inspect.isclass(getattr(self.raw, x)) \
+                             and issubclass(getattr(self.raw, x), ctypes.Structure)}
+            for entry in (_structs - _invalid):
+                yield entry
+            return
+
+    class Modules(dict):
+        def __init__(self, path: pathlib.Path) -> None:
+            self.root: pathlib.Path = path
+
+            for entry in self.root.iterdir():
+                if not entry.is_file(): continue
+                if entry.suffix != ".py": continue
+                if entry.name == "__init__.py": continue
+                module = ExternalStructureManager.Module(entry)
+                self[module.name] = module
+            return
+
+        def __contains__(self, structure_name: str) -> bool:
+            """Return True if the structure name is found in any of the modules"""
+            for module in self.values():
+                if structure_name in module:
+                    return True
+            return False
+
+
+    def __init__(self):
+        self.clear_caches()
+        return
+
+    def clear_caches(self):
+        self._path = None
+        self._modules = None
+        return
+
+    @property
+    def modules(self) -> "ExternalStructureManager.Modules":
+        if not self._modules:
+            self._modules = ExternalStructureManager.Modules(self.path)
+        return self._modules
+
+    @property
+    def path(self) -> pathlib.Path:
+        if not self._path:
+            self._path = pathlib.Path(gef.config["pcustom.struct_path"]).expanduser().absolute()
+        return self._path
+
+    @property
+    def structures(self) -> Generator[Tuple["ExternalStructureManager.Module", "ExternalStructureManager.Structure"], None, None]:
+        for module in self.modules.values():
+            for structure in module.values():
+                yield module, structure
+        return
+
+    @lru_cache()
+    def find(self, structure_name: str) -> Optional[Tuple["ExternalStructureManager.Module", "ExternalStructureManager.Structure"]]:
+        for module in self.modules.values():
+            if structure_name in module:
+                return module, module[structure_name]
+        return None
```

Comment:
It's probably more the role of the caller of `ESL.find` to print those error messages IMO

Reply:
Make sense to raise an exception here for the caller

Reply:
Why isn't `None`  enough?

Reply:
how do we get the error message to the caller? We can totally use `None` as a canary instead.

Reply:
It's a find function, we find what we're looking for (then return the tuple (mod, struct)) or None if not. Exception seems unnecessary.

Reply:
Yes, up to you. Exception would let `find` build the string instead of making each caller build an error string. Or `go` style we could return `value, None` or `None, err`